connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -117,7 +121,38 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
+    responseObserver.onCompleted()
+  }
+
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    import org.apache.arrow.vector.ipc.message.{IpcOption, MessageSerializer}
+
+    val schema = dataframe.schema
+    val maxRecordsPerBatch = dataframe.sparkSession.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = dataframe.sparkSession.sessionState.conf.sessionLocalTimeZone
+    val arrowSchema = ArrowUtils.toArrowSchema(schema, timeZoneId)
+    val schemaBuffer = MessageSerializer.serializeMetadata(arrowSchema, IpcOption.DEFAULT)
+
+    val rows = dataframe.queryExecution.executedPlan.execute().map(_.copy()).collect()
```

Comment:
1, agree, will update
2, i think we can use `toLocalIterator`, but it may trigger multi jobs https://github.com/apache/spark/pull/38300#discussion_r1001301224